### ISSUE TITLE
Varnish not sending INM and IMS headers after 304 from backend

### ISSUE THREAD

**Issue by karptonite** [varnishcache/varnish-cache#2871]
Many apologies if this ought to be a support request--I'm new to this, and it is possible that this is user error.

## Expected Behavior
Varnish should use `If-None-Match` and `If-Modified-Since` where appropriate in requests to the backend when cache is out of ttl and grace but in keep. If seems to fail to do so in the request following a `304 Not Modified` response from the backend.

## Current Behavior
In a pretty simple test case (log below), I have a backend that always returns the same content, and returns a 304 when the INM or IMS headers of the request warrant it. The observation, starting with a completely empty varnish cache, is the following:
First request to the backend caches as normal, keep is set high.
Second request (after it is out of TTL and grace, but in keep): Varnish sends INM and IMS headers, and backend sends 304 not modified with new `Cache-Control` and (identical) `Etag` and `Last-Modified` headers.
Third request, after out of TTL and grace, but in keep: Varnish sends request to backend WITHOUT INM or IMS headers. The backend sends a 200 response.
Fourth request, varnish sends INM and IMS, and toggles back and forth between on subsequent requests.

## Steps to Reproduce (for bugs)
As above, but here are links to the [VCL and Log](https://gist.github.com/karptonite/f99601bd743fe9d33dbf000cc12eeb08)

Notice in the Log: BeReq 6 is the first backend request after cache miss, and it includes the IMS and INM headers. Notice also that the BeResp that it gets back from the server has 304 status, and includes the appropriate Etag and Last-Modified Headers. Then look at BeReq 9, the second miss. It lacks the IMS and INM headers.

## Context
It probably isn't a huge deal, in practice, for us, since the by the time we send the 304, we've done the work to construct the response. But obviously, it would be better if we could send a 304 whenever the cache has not changed, and it is in "keep."

## Your Environment
Varnish 6.0.2
CentOS 7

**Comment by fgsch** [comment]
Broken in addea755c1d41e0431730da666363f76ad00990a. This is unsetting the OF_IMSCAND flag regardless of the response headers.

To complicate things further, by the time we reach this line we have merged the headers from the existing object so we cannot tell whether Last-Modified or Etag was in the original response.

**Comment by fgsch** [comment]
Test:

```
varnishtest ""

server s1 {
	rxreq
	txresp \
		-hdr "last-modified: ${date}" \
		-bodylen 100
	rxreq
	expect req.http.if-modified-since != <undef>
	txresp -status 304
	rxreq
	expect req.http.if-modified-since != <undef>
	txresp -status 304
} -start

varnish v1 -vcl+backend {
	sub vcl_backend_response {
		set beresp.ttl = 1s;
		set beresp.keep = 10s;
	}
} -start

client c1 {
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
	delay 1.5
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
	delay 1.5
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
} -run
```

**Comment by fgsch** [comment]
Test:
```
varnishtest ""

server s1 {
	rxreq
	txresp \
		-hdr {etag: "foo"} \
		-bodylen 100
	rxreq
	expect req.http.if-none-match == {"foo"}
	txresp \
		-status 304 \
		-hdr {etag: "foo"}
	rxreq
	expect req.http.if-none-match == {"foo"}
	txresp \
		-status 304 \
		-hdr {etag: "foo"}
} -start

varnish v1 -vcl+backend {
	sub vcl_backend_response {
		set beresp.ttl = 1s;
		set beresp.keep = 10s;
	}
} -start

client c1 {
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
	delay 1.5
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
	delay 1.5
	txreq
	rxresp
	expect resp.bodylen == 100
	expect resp.status == 200
} -run
```

**Comment by fgsch** [comment]
This is enough to fix this issue and avoid #2763 but we might still want to ensure that either Last-Modified or Etag headers are present in the 304 response for future conditional requests.

```
diff --git a/bin/varnishd/cache/cache_fetch.c b/bin/varnishd/cache/cache_fetch.c
index 7bec80cb9..02ad2cc56 100644
--- a/bin/varnishd/cache/cache_fetch.c
+++ b/bin/varnishd/cache/cache_fetch.c
@@ -648,7 +648,8 @@ vbf_stp_condfetch(struct worker *wrk, struct busyobj *bo)
 		    OA_ESIDATA));
 
 	AZ(ObjCopyAttr(bo->wrk, bo->fetch_objcore, bo->stale_oc, OA_FLAGS));
-	ObjSetFlag(bo->wrk, bo->fetch_objcore, OF_IMSCAND, 0);
+	if (bo->fetch_objcore->flags & OC_F_HFM)
+		ObjSetFlag(bo->wrk, bo->fetch_objcore, OF_IMSCAND, 0);
 	AZ(ObjCopyAttr(bo->wrk, bo->fetch_objcore, bo->stale_oc, OA_GZIPBITS));
 
 	if (bo->do_stream) {
```
